File: mission-to-mars/src/main/java/mars/Sonde.java
package mars;

import java.util.stream.IntStream;

public class Sonde {

	public void preparerRecuperation(ModuleMartien moduleMartien) {
		if (moduleMartien instanceof Helicoptere) {
			poserHelicoptere(moduleMartien);
		}
		moduleMartien.activerRecuperation();
	}

	private void poserHelicoptere(ModuleMartien moduleMartien) {
		int altitude = moduleMartien.getPosition().getZ();
		atterir(moduleMartien, altitude);
	}

	private void atterir(ModuleMartien moduleMartien, int altitude) {
		IntStream.range(0, altitude)
			.forEach((i -> moduleMartien.descendre()));
	}

}

File: typescript/src/Sonde.ts
import { Helicoptere } from './Helicoptere'
import { ModuleMartien } from './ModuleMartien'

export class Sonde {
  preparerRecuperation(moduleMartien: ModuleMartien) {
    if (moduleMartien instanceof Helicoptere) {
      this.poserHelicoptere(moduleMartien);
    }

    moduleMartien.activerRecuperation();
  }

  private poserHelicoptere(moduleMartien: ModuleMartien) {
    const altitude: number = moduleMartien.getPosition().z;

    this.atterir(moduleMartien, altitude);
  }

  private atterir(moduleMartien: ModuleMartien, altitude: number) {
    [...Array(altitude)]
      .forEach(() => { moduleMartien.descendre(); });
  }
}
